### PR TITLE
cis health check, rate limits and domain dev and test case changes

### PR DESCRIPTION
**Pull request by KumarGanesanIBM**

### CODE REVIEW

**Comment by kavya498** [comment]
Please split the pr..
Please have one resource per pr|commit..

**Comment by KumarGanesanIBM** [comment]
> Please split the pr..
> Please have one resource per pr|commit..

it is for only one resource called "CIS Rate Limit" and along with the test case issue in health check and other resource.
It will be very difficult to split it into different PR.  I could not test it if I split it, since it is inter related.
please review it as it is one.
Thanks.

**Comment by KumarGanesanIBM** [comment]
@kavya498 Can you please review this PR?

File: ibm/resource_ibm_cis_healthcheck.go [hunk 1]
```diff
@@ -24,6 +24,7 @@ const (
 	cisGLBHealthCheckAllowInsecure   = "allow_insecure"
 	cisGLBHealthCheckCreatedOn       = "create_on"
 	cisGLBHealthCheckModifiedOn      = "modified_on"
+	cisGLBHealthCheckHeader          = "header"
```
Comment (kavya498) [inline ibm/resource_ibm_cis_healthcheck.go:27]: header is not been included in schema..
header configurations are not been added in resource..

File: ibm/resource_ibm_cis_rate_limit.go [hunk 2]
```diff
@@ -338,220 +338,229 @@ func resourceIBMCISRateLimitValidator() *ResourceValidator {
...
 	if err != nil {
 		return err
 	}
+
 	cisID := d.Get("cis_id").(string)
-	zoneID, _, err := convertTftoCisTwoVar(d.Get("domain_id").(string))
-	if err != nil {
-		return err
-	}
+	zoneID := d.Get("domain_id").(string)
```
Comment (kavya498) [inline ibm/resource_ibm_cis_rate_limit.go:347]: revert back the change..
get zone_id as a combination

File: ibm/resource_ibm_cis_rate_limit.go [hunk 3]
```diff
@@ -145,14 +145,14 @@ func resourceIBMCISRateLimit() *schema.Resource {
...
 				Type:        schema.TypeList,
-				Optional:    true,
+				Required:    true,
 				MaxItems:    1,
 				Description: "Rate Limiting Match",
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"request": {
 							Type:        schema.TypeList,
-							Optional:    true,
```
Comment (kavya498) [inline ibm/resource_ibm_cis_rate_limit.go:155]: revert back the changes
please donot make changes in existing schema

File: ibm/resource_ibm_cis_rate_limit.go [hunk 4]
```diff
@@ -338,220 +338,229 @@ func resourceIBMCISRateLimitValidator() *ResourceValidator {
...
 	return resourceIBMCISRateLimitRead(d, meta)
 }
 
 func resourceIBMCISRateLimitDelete(d *schema.ResourceData, meta interface{}) error {
-	cisClient, err := meta.(ClientSession).CisAPI()
+	cisClient, err := meta.(ClientSession).CisRLClientSession()
 	if err != nil {
 		return err
 	}
-	recordID, zoneID, cisID, _ := convertTfToCisThreeVar(d.Id())
```
Comment (kavya498) [inline ibm/resource_ibm_cis_rate_limit.go:497]: donot do get in delete method,
get it from the id

File: ibm/resource_ibm_cis_rate_limit.go [hunk 5]
```diff
@@ -338,220 +338,229 @@ func resourceIBMCISRateLimitValidator() *ResourceValidator {
...
+	cisClient, err := meta.(ClientSession).CisRLClientSession()
 	if err != nil {
 		return err
 	}
-	updateRecord := v1.RateLimitRecord{}
-	if d.HasChange("disabled") || d.HasChange(cisRLThreshold) || d.HasChange(cisRLPeriod) || d.HasChange(cisRLDescription) || d.HasChange("action") || d.HasChange("match") || d.HasChange("correlate") || d.HasChange("bypass") {
 
-		updateRecord.Threshold = d.Get(cisRLThreshold).(int)
-		updateRecord.Period = d.Get(cisRLPeriod).(int)
+	cisID := d.Get("cis_id").(string)
```
Comment (kavya498) [inline ibm/resource_ibm_cis_rate_limit.go:434]: get it from the id,
do not do d.Get()

File: ibm/resource_ibm_cis_rate_limit_test.go [hunk 6]
```diff
@@ -100,30 +104,35 @@ func testAccCheckIBMCisRateLimitExists(n string, tfRecordID *string) resource.Te
...
 }
 
 func testAccCheckIBMCisRateLimitConfigBasic() string {
-	return testAccCheckIBMCisDomainDataSourceConfig_basic1() + fmt.Sprintf(`
-	
+	return testAccCheckIBMCisDomainDataSourceConfigBasic1() + fmt.Sprintf(`
 	resource "ibm_cis_rate_limit" "ratelimit" {
 		cis_id = data.ibm_cis.cis.id
-		domain_id = data.ibm_cis_domain.cis_domain.id
+		domain_id = data.ibm_cis_domain.cis_domain.domain_id
```
Comment (kavya498) [inline ibm/resource_ibm_cis_rate_limit_test.go:135]: revert back to `domain_id = data.ibm_cis_domain.cis_domain.id'
 for all existing resource testcases

File: ibm/data_source_ibm_cis_rate_limit.go [hunk 7]
```diff
@@ -178,38 +179,40 @@ func dataSourceIBMCISRateLimit() *schema.Resource {
 }
 
 func dataSourceIBMCISRateLimitRead(d *schema.ResourceData, meta interface{}) error {
-	cisClient, err := meta.(ClientSession).CisAPI()
+	cisClient, err := meta.(ClientSession).CisRLClientSession()
 	if err != nil {
 		return err
 	}
 	cisID := d.Get("cis_id").(string)
-	zoneID, _, err := convertTftoCisTwoVar(d.Get("domain_id").(string))
```
Comment (kavya498) [inline ibm/data_source_ibm_cis_rate_limit.go:186]: revert this change in the existing datasource

File: ibm/resource_ibm_cis_healthcheck.go (hunk 1 above)
Comment (KumarGanesanIBM) [reply ibm/resource_ibm_cis_healthcheck.go]: I have removed this line. I will commit this feature in another PR.

File: ibm/resource_ibm_cis_rate_limit.go (hunk 2 above)
Comment (KumarGanesanIBM) [reply ibm/resource_ibm_cis_rate_limit.go]: done

File: ibm/resource_ibm_cis_rate_limit.go (hunk 3 above)
Comment (KumarGanesanIBM) [reply ibm/resource_ibm_cis_rate_limit.go]: done

Comment (KumarGanesanIBM) [reply ibm/resource_ibm_cis_rate_limit.go]: But, "request" property is a mandatory field to rate limit. if user did not specify request property in tf code, they will get a crash log. 

I made it optional, since, you have asked to make it optional.

File: ibm/resource_ibm_cis_rate_limit.go (hunk 4 above)
Comment (KumarGanesanIBM) [reply ibm/resource_ibm_cis_rate_limit.go:497]: done

File: ibm/resource_ibm_cis_rate_limit.go (hunk 5 above)
Comment (KumarGanesanIBM) [reply ibm/resource_ibm_cis_rate_limit.go]: done

File: ibm/data_source_ibm_cis_rate_limit.go (hunk 7 above)
Comment (KumarGanesanIBM) [reply ibm/data_source_ibm_cis_rate_limit.go]: done

File: ibm/resource_ibm_cis_rate_limit.go (hunk 3 above)
Comment (KumarGanesanIBM) [reply ibm/resource_ibm_cis_rate_limit.go:155]: made the request field as optional and computed.
modified the expandMatch code to set the default match request value.
Now it works if we don't specify match and request property.

**Comment by KumarGanesanIBM** [comment]
@hkantare @kavya498 
I am not able to raise PR for other resource, since there are dependencies for this resource code.  @kavya498 It'd be great if you review it as soon as possible.
Thanks...
